service/cziti/dns-proxy.go

```diff
@@ -350,32 +350,36 @@ outer:
 
 	reqs := make(map[uint32]*proxiedReq)
 
-	log.Debug("Upstream DNS proxy loop begins")
-	for {
-		select {
-		case pr := <-proxiedRequests:
-			id := (uint32(pr.req.Id) << 16) | uint32(pr.req.Question[0].Qtype)
-			reqs[id] = pr
-			b, _ := pr.req.Pack()
-			for _, proxy := range dnsUpstreams {
-				if _, err := proxy.Write(b); err != nil {
-
-					_ = proxy.Close() //first thing - close the proxy connection
-
-					// when this hits - it never seems to recover. throw a panic which will get recovered
-					// via the defer specified above and try to recreate the connections. this is a heavy
-					// handed way of reconnecting to the DNS server but it should be infrequent
-					log.Panicf("failed to proxy DNS to %s %s %v. %v",
-						dns.Type(pr.req.Question[0].Qtype),
-						pr.req.Question[0].Name,
-						proxy.RemoteAddr(),
-						err)
-				} else {
-					log.Tracef("Proxied request sent to %v from ipv%d listener", proxy.RemoteAddr(), pr.ipVer)
+	tickerChannel := make(chan bool)
+	ticker := time.NewTicker(60 * time.Second)
+	defer func() {
+		tickerChannel <- false
+		log.Tracef("Removing timer tasks for dns request handling")
+	}()
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				// cleanup requests we didn't get answers for
+				now := time.Now()
+				for k, r := range reqs {
+					if now.After(r.exp) {
+						log.Debugf("a DNS request has expired - enable trace logging and reproduce this issue for more information")
+						log.Tracef("expired DNS req: %s %s", dns.Type(r.req.Question[0].Qtype), r.req.Question[0].Name)
+
+						delete(reqs, k)
```

Comment:
i might be wrong but i am pretty sure using the reqs map will no longer be threadsafe when it's in go routines like this, right?

Reply:
Yes, I missed it. I will add a lock before updating the map.

Reply:
i think there's another problem cleaning the go-routines on 380 when the panic on new line 418 triggers due to network events. it will continue to run and might cause trouble. If nothing else it'd be in that for loop forever.